File: 64.最小路径和.cpp
// dfs超时
class Solution {
public:
    int dis[2][2] = {{1, 0}, {0, 1}};
    int minSum = INT_MAX;

    int minPathSum(vector<vector<int>>& grid) {
        if(grid.size() == 0)
            return 0;
        int m = grid.size();
        int n = grid[0].size();
        dfs(0, 0, m, n, grid, 0);
        return minSum;
    }

   void  dfs(int r, int c, int m, int n, vector<vector<int>>& grid, int sum){
       sum += grid[r][c];
       if(r == m - 1 && c == n - 1){
           minSum = min(minSum, sum);
           return;
       }
       for(int i = 0; i < 2; ++i){
            int row = r + dis[i][0];
            int col = c + dis[i][1];
            if(row >= m || col >= n)
                continue;
            dfs(row, col, m, n, grid, sum);
       }
   }
};

// dp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        if(m == 0 || n == 0)
            return 0;
        // 初始化
        for(int i = 1; i < m; ++i)
            grid[i][0] += grid[i - 1][0];
        for(int i = 1; i < n; ++i)
            grid[0][i] += grid[0][i - 1];
        for(int i = 1; i < m; ++i)
            for(int j = 1; j < n; ++j)
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1]);
        return grid[m - 1][n - 1];
    }
};

File: LeetCode 热题 HOT 100/64.最小路径和.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
    let m = grid.length,
        n = grid[0].length;
    for (let i = 1; i < m; ++i)
        grid[i][0] += grid[i - 1][0];
    for (let j = 1; j < n; ++j)
        grid[0][j] += grid[0][j - 1];
    for (let i = 1; i < m; ++i) {
        for (let j = 1; j < n; ++j) {
            grid[i][j] += Math.min(grid[i - 1][j], grid[i][j - 1]);
        }
    }
    return grid[m - 1][n - 1];
};

// dfs 超时
var minPathSum = function (grid) {
    let m = grid.length,
        n = grid[0].length;
    let res = Number.MAX_SAFE_INTEGER;

    const dfs = (i, j, m, n, total) => {
        if (i === m - 1 && j === n - 1) {
            res = Math.min(res, total + grid[i][j]);
            return;
        }
        if (j < n - 1) {
            dfs(i, j + 1, m, n, total + grid[i][j]);
        }
        if (i < m - 1) {
            dfs(i + 1, j, m, n, total + grid[i][j]);
        }
    }

    dfs(0, 0, m, n, 0);
    return res;
};